ocaml/xapi/xapi_pgpu_helpers.ml

```diff
@@ -136,82 +137,108 @@ let assert_capacity_exists_for_VGPU_type ~__context ~self ~vgpu_type =
+    | Some (rpc, session_id) -> XenAPI.PGPU.get_GPU_group rpc session_id pgpu
   in
-  let get_compatibility_metadata pgpu =
+  let get_gpu_types_of_gpu_group gpu_group =
     match remote with
-    | None -> Db.PGPU.get_compatibility_metadata ~__context ~self:pgpu
-    | Some (rpc, session_id) ->XenAPI.PGPU.get_compatibility_metadata rpc session_id pgpu
+    | None -> Db.GPU_group.get_GPU_types ~__context ~self:gpu_group
+    | Some (rpc, session_id) -> XenAPI.GPU_group.get_GPU_types rpc session_id gpu_group
   in
-  (* extract vgpu implementation *)
-  let vgpu_impl vgpu = 
-    vgpu
-    |> (fun self -> Db.VGPU.get_type ~__context ~self)
-    |> (fun self -> Db.VGPU_type.get_implementation ~__context ~self)
+  let get_pgpus_of_gpu_group gpu_group =
+    match remote with
+    | None -> Db.GPU_group.get_PGPUs ~__context ~self:gpu_group
+    | Some (rpc, session_id) -> XenAPI.GPU_group.get_PGPUs rpc session_id gpu_group
   in
   (* get first pgpu with the correct implementation, the vgpu is needed for the error metadata *)
-  let rec get_first_pgpu_of_implementation vgpu implementation = function
-    | [] -> 
+  let rec get_first_suitable_pgpu pgpu_types vgpu = function
+    | [] ->
       debug "No compatible pGPU found on host %s, needed by VM %s" (Ref.string_of host) (Ref.string_of vm);
       let pgpu_group = Db.VGPU.get_GPU_group ~__context ~self:vgpu in
       raise Api_errors.(Server_error (vm_requires_gpu, [Ref.string_of vm; Ref.string_of pgpu_group]))
-    | pgpu :: rest -> 
-      let supported = 
-        get_supported_vgpu_types pgpu
-        |> List.exists (has_implementation implementation)
+    | pgpu :: rest ->
+      let types =
+        get_gpu_group_of_pgpu pgpu
+        |> get_gpu_types_of_gpu_group
       in
-      if supported then pgpu
-      else get_first_pgpu_of_implementation vgpu implementation rest
-  in
-  let test_nvidia_compatibility vgpu pgpu = 
-    let pgpu_metadata =
-      let key = Xapi_gpumon.Nvidia.key in
-      try
-        get_compatibility_metadata pgpu
-        |> List.assoc key
-      with Not_found -> 
-        debug "Key %s is missing from the compatibility_metadata for pgpu %s on the host %s." key (Ref.string_of pgpu) (Ref.string_of host);
-        raise Api_errors.(Server_error (nvidia_tools_error, [Ref.string_of host]))
-    in
-    Xapi_gpumon.Nvidia.assert_pgpu_is_compatible_with_vm ~__context ~vm ~vgpu ~dest_host:host ~encoded_pgpu_metadata:pgpu_metadata
+      if Listext.List.set_equiv pgpu_types types
+      then pgpu
+      else get_first_suitable_pgpu pgpu_types vgpu rest
   in
   let test_compatibility vgpu pgpus =
-    match vgpu_impl vgpu with
-      | `passthrough | `gvt_g | `mxgpu -> ()
-      | `nvidia as impl -> 
-        let pgpu = get_first_pgpu_of_implementation vgpu impl pgpus
-        in test_nvidia_compatibility vgpu pgpu
+    match vgpu_impl ~__context vgpu with
+    | `passthrough | `gvt_g | `mxgpu -> ()
+    | `nvidia ->
+      let pgpu_types =
```

Comment:
Not sure this is an improvement. I tend to write pipelines when there is a clear transformation and none of the intermediate value is required down below.
```
Db.VGPU.get_GPU_group ~__context ~self:vgpu
|> fun self -> Db.GPU_group.get_GPU_types ~__context ~self
|> fun pgpu_types -> get_first_suitable_pgpu pgpu_types vgpu pgpus
|> fun pgpu -> assert_destination_pgpu_is_compatible_with_vm ~__context ~vm ~vgpu ~pgpu ~host ?remote ()
```